3_cluster/src/seasonal_metric_cluster.R

```diff
@@ -0,0 +1,275 @@
+#Function to compute the clusters for seasonal metrics
+# written to be branched over the metric
+#metric_mat - p1_FDC_metrics_season. rows are gauges, columns are seasonal metrics
+#metric - the name of the metric without the _s season at the end
+#dist_method - the distance computation for dist()
+seasonal_metric_cluster <- function(metric_mat, metric, 
+                                    dist_method = 'euclidean'
+                                    ){
+  #Select all of the seasonal columns for this metric
+  metric_mat <- metric_mat[, c(1,grep(x = colnames(metric_mat), pattern = paste0(metric,'_')))]
+  
+  #Scaling of metrics should not be necessary because the metrics are on [0,1]
+  
+  #Compute the distance matrix between all sites
+  dists <- dist(metric_mat[,-1], method = dist_method)
+  
+  #Compute clusters using different methods
+  clust_methods_hclust <- c( "average", "single", "complete", "ward.D2")
+  names(clust_methods_hclust) <- clust_methods_hclust
+  
+  #compute clusters using hclust function instead of agnes becuase it's faster
+  clusts <- purrr::map(.x = clust_methods_hclust,
+                      .f = hclust, 
+                      d = dists, members = NULL)
+  #add metric name
+  #add agglomeration coefficient
+  for (i in 1:length(clusts)){
+    clusts[[i]]$ac = round(coef.hclust(clusts[[i]]), 3)
+    clusts[[i]]$metric = metric
+  }
+  
+  return(clusts)
+}
+
+#Function to extract the best cluster method for each metric
+#clusts is the list output from seasonal_metric_cluster
+select_cluster_method <- function(clusts){
+  #data.frame of the metric, method, and ac value
+  df <- matrix(nrow = length(clusts), ncol = 3, data = '')
+  for (i in 1:nrow(df)){
+    df[i,] <- as.character(clusts[[i]][c('metric', 'method', 'ac')])
+  }
+  df <- as.data.frame(df)
+  colnames(df) <- c('metric', 'method', 'ac')
+  df$ac <- as.numeric(df$ac)
+  
+  #df of the clustering method with the max ac value for each metric
+  df_max <- df[1:length(unique(df$metric)),]
+  df_max[,] <- NA
+  for (i in 1:nrow(df_max)){
+    a <- df[df$metric == unique(df$metric)[i],]
+    df_max[i,] <- a[a$ac == max(a$ac),]
+  }
+  
+  return(df_max)
+}
+
+#Function to make cluster diagnostic panel plot
+plot_cluster_diagnostics <- function(clusts, metric_mat, 
+                                     dist_method = 'euclidean',
```

Comment:
How are we going to choose the optimal number of clusters? Are they going to be based on these diagnostic plots in some automated way?  Looking at some of the plots, the place where it draws a vertical line for gap statistic seems a little off sometimes. I guess it just chooses the maximum value, but sometimes when the values plateau and then gradually increase, it just puts that line way out at 20 or something when probably it should be closer to the inflection point.  Anyway - just something to think about -  if we are going to be choosing our final number of clusters from these then we probably should make sure we agree with how those vertical lines are drawn.

Reply:
The vertical lines in the gap statistic plots are currently set to take the global maximum, which must be based on the bootstrapped mean value and not consider the uncertainty. All of the built-in methods for drawing those vertical optimality lines do not place them where I would place them considering the uncertainty, as you point out. I think the range of optimality is from maybe 8-15 clusters for most metrics. There are many other statistics that can be evaluated to assess the optimal number of clusters (30 in NbClust) and I think a histogram of the optimal number from those 30 metrics would be helpful to justify our final choice. I can add making that histogram as an issue if that sounds good.